service/controller/app/v1/resource/cordonchart/resource.go

```diff
@@ -0,0 +1,139 @@
+type Config struct {
+	// Dependencies.
+	Logger micrologger.Logger
+
+	// Settings.
+	ChartNamespace string
+}
+
+type patchSpec struct {
+	Op    string      `json:"op"`
+	Path  string      `json:"path"`
+	Value interface{} `json:"value,omitempty"`
+}
+
+// Resource implements the cordonchart resource.
+type Resource struct {
+	// Dependencies.
+	logger micrologger.Logger
+
+	// Settings.
+	chartNamespace string
+}
+
+// New creates a new configured cordonchart resource.
+func New(config Config) (*Resource, error) {
+	if config.Logger == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.Logger must not be empty", config)
+	}
+
+	if config.ChartNamespace == "" {
+		return nil, microerror.Maskf(invalidConfigError, "%T.ChartNamespace must not be empty", config)
+	}
+
+	r := &Resource{
+		logger:         config.Logger,
+		chartNamespace: config.ChartNamespace,
+	}
+
+	return r, nil
+}
+
+func (r Resource) Name() string {
+	return Name
+}
+
+func (r *Resource) addCordon(ctx context.Context, cr v1alpha1.App, client versioned.Interface) error {
+	var err error
+	var patchByte []byte
+	{
+		patch := []patchSpec{
+			{
+				Op:    "add",
+				Path:  replaceToEscape(fmt.Sprintf("/metadata/annotations/%s", replacePrefix(annotation.CordonUntil))),
+				Value: key.CordonUntil(cr),
+			},
+			{
+				Op:    "add",
+				Path:  replaceToEscape(fmt.Sprintf("/metadata/annotations/%s", replacePrefix(annotation.CordonReason))),
+				Value: key.CordonReason(cr),
+			},
+		}
+
+		patchByte, err = json.Marshal(patch)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+	}
+
+	_, err = client.ApplicationV1alpha1().Charts(r.chartNamespace).Patch(cr.GetName(), types.JSONPatchType, patchByte)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+	r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("added cordon annotations for chart CR %#q in namespace %#q", cr.Name, r.chartNamespace))
+
+	return nil
+}
+
+func (r *Resource) deleteCordon(ctx context.Context, cr v1alpha1.App, client versioned.Interface) error {
+	var err error
+	var patchByte []byte
+
+	{
+		patch := []patchSpec{
+			{
+				Op:   "remove",
+				Path: replaceToEscape(fmt.Sprintf("/metadata/annotations/%s", replacePrefix(annotation.CordonUntil))),
+			},
+			{
+				Op:   "remove",
+				Path: replaceToEscape(fmt.Sprintf("/metadata/annotations/%s", replacePrefix(annotation.CordonReason))),
+			},
+		}
+
+		patchByte, err = json.Marshal(patch)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+	}
+
+	_, err = client.ApplicationV1alpha1().Charts(r.chartNamespace).Patch(cr.GetName(), types.JSONPatchType, patchByte)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+	r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("deleted cordon annotations for chart CR %#q in namespace %#q", cr.Name, r.chartNamespace))
+
+	return nil
+}
+
+func replacePrefix(from string) string {
+	return strings.Replace(from, "app-operator.giantswarm.io", "chart-operator.giantswarm.io", 1)
```

Comment:
I feel like this is done in a lot of different places. How can we make this a bit better? Why aren't we using the label package here at least? I am also still wondering why we need to do this at all. We patch a chart CR. So why does the chart CR have an `app-operator` annotation at all?

Reply:
@xh3b4sd If we are moving this function to another package, I think it should be `annotations`. 
And this function is to convert `app-operator.giantswarm.io` prefix into `chart-operator.giantswarm.io` so `chart` CR would have no `app-operator` annotations when the operator is done with copying these.

Reply:
My question was why the chart CR still haves the `app-operator` annotations when it got copied and why we do this at all. My question was if you can explain the whole thing to me a bit more.

Reply:
Also note that I did not say we should move this function to the annotation package. I asked the question how we could improve this and how it should be so that we have a plan or idea or vision for this.

Reply:
Annotations related to cordon features are from `app` CRs, and we copy the same cordon annotations into `chart` CRs in TC to stop affecting into helm charts. But only we need to change the prefix in annotations so hiding the information the app-operator is manipulating these cordon job from the background. 
Also, Having `chart` CR with other operator's prefix is wrong in the design concept I believe. 

That's why I thought about removing this prefix for this cordon annotation once for all if we need to copy this information from CRs to CRs in another type. 
What makes sense to have prefix from the start in this case?
@xh3b4sd WDYT?

Reply:
Who decides when to cordon an app? How does this process look like? I am wondering why two different CRs and two different operator responsibilities need to be involved here. This copy back and forth is something I do not get into my head. I do not understand that. Why do we need the cordon annotations in two different places?

Reply:
`chart` CRs in TC are merely the reconciliation result from `app` CR, therefore, we need to set annotations into `app` CRs to have it in `chart` CRs. That's why we involved two different operators in here. 
If you have more questions I think the technical spec for phase 2 would be very helpful. 
https://github.com/giantswarm/giantswarm/blob/master/specs/app-catalog-phase-2.md
Or we can jump into a joint call about this.

Reply:
Thank you for the spec link. I start to understand it. What the spec does not answer apparently is why we cordon at the start of the responsibility chain instead of at the end. Why don't we "ignore" `app-operator` in that regard and only cordon the `Chart` CRs in the TCs?

Reply:
@xh3b4sd What if we trying to change the user config or chart version inside `app` CR? I think we need to mitigate those changes before it affects `chart` CRs in TC. So far only `cordon` related annotation can stop this, WDYT?